Documentation added for testing (#1281)
NVIDIA/spark-rapids: docs/testing.md

## Unit tests

> Unit tests exist in the [here](tests) directory. This is unconventional and is done so we can run

Reviewer: "unit tests exist in the here directory" is not very readable.  Also not sure why you are linking to.  There is no tests directory or file under docs that this can or should link to.

>  run the tests on the final shaded version of the plugin. It also helps with how we collect code coverage.

Reviewer 2: ```suggestion
run the tests on the final shaded version of the plugin. It also helps with how we collect code coverage.
```

> in order to run the unit-tests follow these steps

Reviewer: Please capitalize sentences.

> In order to run the unit-tests issue the maven command to run the tests i.e. `mvn test`. 

Reviewer 2: ```suggestion
Use Maven to run the unit tests via `mvn test`. 
```

> To run individual tests append `-DwildcardSuites=<comma separated list of wildcard suite names to execute>` to the above command 

Reviewer 3: does this work for both scala and java tests? I know the java ones are different in spark

Reviewer 2: wildcardSuites only applies to scalatest.  Java tests need to be filtered via surefire's approach which is `-Dtest=...`

Reviewer 3: right so we have a java test right - TestHashedPriorityQueue.java?  so adding how to run it would be useful and specifying here this works for scala tests.

Reviewer 3: I guess this text is old new one mentions scala tests

> 2. to run individual tests append `-dwildcardsuites=<comma separated list of wildcard suite names to execute>` to the above command 

Reviewer: These are not steps.  They are options and I don't think we need a list to express them.

> For more information about using scalatest with maven please refer [here](https://www.scalatest.org/user_guide/using_the_scalatest_maven_plugin)

Reviewer 2: ```suggestion
For more information about using scalatest with Maven please refer to the
[scalatest documentation](https://www.scalatest.org/user_guide/using_the_scalatest_maven_plugin).
```

> for more information about using scalatest with maven please refere [here](https://www.scalatest.org/user_guide/using_the_scalatest_maven_plugin)

Reviewer: I'm not known for my spelling but I don't think refer has an "e" at the end

    
#### Running unit-tests against specific apache spark versions. 
You can run the unit tests against different versions of Spark using the different profiles. The
default version runs against Spark 3.0.0, to run against other versions use one of the following
 profiles:
   - `-Pspark301tests` (spark 3.0.1)
   - `-Pspark302tests` (spark 3.0.2)
   - `-Pspark310tests` (spark 3.1.0)

> Please refer to the [pom.xml](tests/pom.xml) to see the list of profiles supported
> apache spark specific configurations can be passed in by setting environment-variable spark_conf 

Reviewer 2: ```suggestion
Please refer to the [tests project POM](tests/pom.xml) to see the list of test profiles supported.
Apache Spark specific configurations can be passed in by setting the `SPARK_CONF` environment
variable.
```

Examples: 

> -To run tests against Apache Spark 3.1.0, 

Reviewer 2: This says it will run tests against Spark 3.1.0 but only runs one test.  Either fix the description or fix the example.

 `mvn -P spark310tests test -DwildcardSuites="com.nvidia.spark.rapids.ParquetWriterSuite"`.
- To pass Apache Spark configs `--conf spark.dynamicAllocation.enabled=false --conf spark.task.cpus=1` do something like.
 `SPARK_CONF="spark.dynamicAllocation.enabled=false,spark.task.cpus=1" mvn ...`
- To run test ParquetWriterSuite in package com.nvidia.spark.rapids, issue `mvn test -DwildcardSuites="com.nvidia.spark.rapids.ParquetWriterSuite"`

## Integration tests

> Integration tests are stored in the [integration_tests](integration_tests) directory. 

Reviewer: This is not in the right directory for this link to work.  Please change it back to the README or at least ../integration_tests

There are two frameworks used for testing. one is based off of pytest and pyspark in the 

> `src/main/python` directory. These tests will run as a part of the maven build if you have the 

Reviewer 2: ```suggestion
`src/main/python` directory. These tests will run as a part of the Maven build if you have the 
```

environment variable `SPARK_HOME` set.

The suggested way to run these tests is to use the shell-script file located in the module folder 
called [run_pyspark_from_build.sh](integration_tests/run_pyspark_from_build.sh). This script takes 
care of some of the flags that are required to run the tests which will have to be set for the 
plugin to work. it will be very useful to read the contents of the 
[run_pyspark_from_build.sh](integration_tests/run_pyspark_from_build.sh) to get a better insight 
into what is needed as we constantly keep working on to improve and expand the plugin-support.

The tests are written python and run with pytest and the script honors pytest parameters. Some handy flags are:
- `-k` <pytest-file-name>. This will run all the tests in that test file.
- `-k` <test-name>. This will also run an individual test.
- `-s` Doesn't capture the output and instead prints to the screen.
- `-v` Increase the verbosity of the tests

> - `-rfexxs` Show extra test summary info as specified by chars: (f)ailed, (e)rror, (x)failed, (x)passed, (s)kipped

Reviewer 2: These flags are wrong.  The first clue is the same character meaning two different options.  It should be something like this:
```suggestion
- `-r fExXs` Show extra test summary info as specified by chars: (f)ailed, (E)rror, (x)failed, (X)passed, (s)kipped
```

durations.run(new com.nvidia.spark.rapids.JoinsSuite)
...
```

> For more details on Integration tests options please refer to integration-test [doc](integration-tests/README.md)

Reviewer: Need to test out the links I don't think this works.
